src/ConnectedMode.UnitTests/QualityProfiles/QualityProfileDownloaderTests.cs

```diff
@@ -0,0 +1,320 @@
+
+            solutionBindingOperation.Invocations.Count().Should().Be(1);
+            var savedConfigs = solutionBindingOperation.Invocations[0].Arguments[0] as IEnumerable<IBindingConfig>;
+
+            savedConfigs.Should().BeEquivalentTo(csharpConfig, vbnetConfig);
+
+            // Check output messages
+            var missingPluginMessageCFamily = string.Format(Strings.SubTextPaddingFormat,
+                string.Format(BindingStrings.CannotDownloadQualityProfileForLanguage, Language.Cpp.Name));
+            var missingPluginMessageSecrets = string.Format(Strings.SubTextPaddingFormat,
+                string.Format(BindingStrings.CannotDownloadQualityProfileForLanguage, Language.Secrets.Name));
+
+            logger.AssertOutputStringExists(missingPluginMessageCFamily);
+            logger.AssertOutputStringExists(missingPluginMessageSecrets);
+            logger.AssertPartialOutputStringDoesNotExist(Language.Css.Name);
+        }
+
+        [TestMethod]
+        public async Task UpdateAsync_WhenBindingConfigIsNull_Fails()
+        {
+            // Arrange
+            var logger = new TestLogger(logToConsole: true);
+
+            var language = Language.VBNET;
+            var sonarQubeService = new Mock<ISonarQubeService>();
+            SonarQubeQualityProfile profile = ConfigureQualityProfile(sonarQubeService, language);
+
+            var solutionBindingOperation = new Mock<ISolutionBindingOperation>();
+            var boundProject = CreateBoundProject("key", "any project name", new Uri("http://connected"));
+            var testSubject = CreateTestSubject(
+                languagesToBind: new[] { language },
+                sonarQubeService: sonarQubeService.Object,
+                solutionBindingOperation: solutionBindingOperation.Object,
+                logger: logger);
+
+            // Act
+            var result = await testSubject.UpdateAsync(boundProject, null, CancellationToken.None);
+
+            // Assert
+            result.Should().BeFalse();
+            solutionBindingOperation.Invocations.Should().BeEmpty();
+
+            logger.AssertOutputStrings(1);
+            var expectedOutput = string.Format(Strings.SubTextPaddingFormat,
+                string.Format(BindingStrings.FailedToCreateBindingConfigForLanguage, language.Name));
+            logger.AssertOutputStrings(expectedOutput);
+        }
+
+        [TestMethod]
+        public async Task UpdateAsync_SavesConfiguration()
+        {
+            var configPersister = new DummyConfigPersister();
+
+            // Arrange
+            var bindingConfig = new Mock<IBindingConfig>().Object;
+
+            var language = Language.VBNET;
+            var serverQpTimestamp = DateTime.UtcNow.AddHours(-2);
+            var sonarQubeService = new Mock<ISonarQubeService>();
+            SonarQubeQualityProfile profile = ConfigureQualityProfile(sonarQubeService, language,
+                "my profile key", serverQpTimestamp);
```

Comment:
Pre-existing test, moved from the other test class.
I tweaked it slightly to explicitly set the expected profile name and profile timestamp to specific values.